### PR TITLE
Allow downloading requirements file from GCS in `BeamRunPythonPipelineOperator`

### PR DESCRIPTION
**Pull request by yupbank**
currently only python file support gcs download, however, requirement.txt might also benefit from that

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/apache/beam/operators/beam.py [hunk 1]
```diff
@@ -294,10 +294,15 @@ def execute(self, context: Context):
...
-            if self.py_file.lower().startswith("gs://"):
+                
+            if self.py_file.lower().startswith("gs://") or snake_case_pipeline_options.get('requirements_file', '').startswith("gs://"):
                 gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
                 tmp_gcs_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=self.py_file))
                 self.py_file = tmp_gcs_file.name
+                if snake_case_pipeline_options.get('requirements_file', '').startswith("gs://"):
+                    tmp_req_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=snake_case_pipeline_options['requirements_file']))
+                    snake_case_pipeline_options['requirements_file'] = tmp_req_file.name                              
+
```
Comment (hussein-awala) [inline airflow/providers/apache/beam/operators/beam.py]: can we have a py_file which doesn't start with `gs://` with a requirement file stored in gcs in the same job? if yes (and I think it's yes), your code will try to download the two files from gcs.

Instead, you can create a method to download an object from gcs, and use it for py_file and/or requirements file:
```python
def _get_file_from_gcs(self, object_url):
    gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
    return exit_stack.enter_context(gcs_hook.provide_file(object_url=object_url))
```
and in this method:
```python
if self.py_file.lower().startswith("gs://"):
    self.py_file = self._get_file_from_gcs(self.py_file).name
if snake_case_pipeline_options.get('requirements_file', '').startswith("gs://"):
    snake_case_pipeline_options['requirements_file'] = self._get_file_from_gcs(snake_case_pipeline_options['requirements_file']).name
```
WDYT?

**Review by hussein-awala — COMMENTED** [review]
don't forget to add a test for your change

File: airflow/providers/apache/beam/operators/beam.py (hunk 1 above)
Comment (yupbank) [reply airflow/providers/apache/beam/operators/beam.py]: sounds good, i'll do that instead

File: airflow/providers/apache/beam/operators/beam.py [hunk 2]
```diff
@@ -294,10 +294,13 @@ def execute(self, context: Context):
...
 
         with ExitStack() as exit_stack:
+            gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
             if self.py_file.lower().startswith("gs://"):
-                gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
                 tmp_gcs_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=self.py_file))
                 self.py_file = tmp_gcs_file.name
+            if snake_case_pipeline_options.get('requirements_file', '').startswith("gs://"):
+                tmp_req_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=snake_case_pipeline_options['requirements_file']))
+                snake_case_pipeline_options['requirements_file'] = tmp_req_file.name
```
Comment (hussein-awala) [inline airflow/providers/apache/beam/operators/beam.py]:
```suggestion
            if snake_case_pipeline_options.get("requirements_file", "").startswith("gs://"):
                tmp_req_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=snake_case_pipeline_options["requirements_file"]))
                snake_case_pipeline_options["requirements_file"] = tmp_req_file.name
```

**Review by hussein-awala — APPROVED** [review]

File: airflow/providers/apache/beam/operators/beam.py [hunk 3]
```diff
@@ -294,10 +294,13 @@ def execute(self, context: Context):
             raise AirflowException("Beam hook is not defined.")
 
         with ExitStack() as exit_stack:
+            gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
             if self.py_file.lower().startswith("gs://"):
-                gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
                 tmp_gcs_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=self.py_file))
                 self.py_file = tmp_gcs_file.name
+            if snake_case_pipeline_options.get("requirements_file", "").startswith("gs://"):
+                tmp_req_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=snake_case_pipeline_options["requirements_file"]))
```
Comment (hussein-awala) [inline airflow/providers/apache/beam/operators/beam.py]:
```suggestion
                tmp_req_file = exit_stack.enter_context(
                    gcs_hook.provide_file(object_url=snake_case_pipeline_options["requirements_file"])
                 )
```

File: airflow/providers/apache/beam/operators/beam.py [hunk 4]
```diff
@@ -294,10 +294,15 @@ def execute(self, context: Context):
...
         with ExitStack() as exit_stack:
+            gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
             if self.py_file.lower().startswith("gs://"):
-                gcs_hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
                 tmp_gcs_file = exit_stack.enter_context(gcs_hook.provide_file(object_url=self.py_file))
                 self.py_file = tmp_gcs_file.name
+            if snake_case_pipeline_options.get("requirements_file", "").startswith("gs://"):
+                tmp_req_file = exit_stack.enter_context(
+                    gcs_hook.provide_file(object_url=snake_case_pipeline_options["requirements_file"])
+                 )
```
Comment (hussein-awala) [inline airflow/providers/apache/beam/operators/beam.py]:
```suggestion
                )
```

**Comment by hussein-awala** [comment]
@yupbank could you check the failed tests?

**Review by potiuk — APPROVED** [review]

**Pull request merged**